### PR TITLE
Db expression

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️/❌
| Tests pass?   | ✔️
| Fixed issues  | #18134 

 Expression is not will be quoted

### CODE REVIEW

File: framework/db/Connection.php [hunk 1]
```diff
@@ -925,11 +925,16 @@ public function quoteTableName($name)
      * If the column name contains prefix, the prefix will also be properly quoted.
      * If the column name is already quoted or contains special characters including '(', '[[' and '{{',
      * then this method will do nothing.
+     * If the column name is Exresion - return as is
```
Comment (samdark) [inline framework/db/Connection.php:928]:
```suggestion
     * If the column name is an instance of `ExpressionInterface` it is returned as is.
```

File: framework/db/Connection.php [hunk 2]
```diff
@@ -925,11 +925,16 @@ public function quoteTableName($name)
...
      * @param string $name column name
      * @return string the properly quoted column name
      */
     public function quoteColumnName($name)
     {
+        // Expression is not will be quoted (https://github.com/yiisoft/yii2/issues/18134)
```
Comment (samdark) [inline framework/db/Connection.php:934]:
```suggestion
```

File: framework/db/Connection.php [hunk 3]
```diff
@@ -925,11 +925,16 @@ public function quoteTableName($name)
      * If the column name contains prefix, the prefix will also be properly quoted.
      * If the column name is already quoted or contains special characters including '(', '[[' and '{{',
      * then this method will do nothing.
+     * If the column name is Exresion - return as is
      * @param string $name column name
```
Comment (samdark) [inline framework/db/Connection.php:929]:
```suggestion
     * @param string $name|ExpressionInterface column name
```

File: tests/framework/db/QueryBuilderTest.php [hunk 4]
```diff
@@ -2523,6 +2523,17 @@ public function testBuildLikeCondition($condition, $expected, $expectedParams)
         $this->assertEquals($expectedParams, $params);
     }
 
+    /**
+     * @see https://github.com/yiisoft/yii2/issues/18134
+     */
+    public function testIssue18134()
```
Comment (samdark) [inline tests/framework/db/QueryBuilderTest.php:2529]:
```suggestion
    public function testExpressionIsNotQuotedInColumnName()
```

**Review by samdark — CHANGES_REQUESTED** [review]
Overall it's fine. Need some minor things fixed and a line for CHANGELOG.md

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Review by rob006 — COMMENTED** [review]
I don't think that this is right place to fix it. Expressions should be handled at `LikeConditionBuilder::build()` level, there is no need to add special conditions in `Connection::quoteColumnName()`.

See https://github.com/yiisoft/yii2/pull/17490

**Comment by darkdef** [comment]
But what about the case when the Expression will be used instead of the column name in other cases?
This issue (for example): https://github.com/yiisoft/yii2/issues/17473

File: framework/db/Connection.php [hunk 5]
```diff
@@ -925,11 +925,15 @@ public function quoteTableName($name)
      * If the column name contains prefix, the prefix will also be properly quoted.
      * If the column name is already quoted or contains special characters including '(', '[[' and '{{',
      * then this method will do nothing.
-     * @param string $name column name
+     * If the column name is an instance of `ExpressionInterface` it is returned as is.
+     * @param string $name|ExpressionInterface column name
```
[the remark this answers is not in the document]
Comment (samdark) [reply framework/db/Connection.php:929]: Why?

[the remark this answers is not in the document]
Comment (darkdef) [reply framework/db/Connection.php:929]: @param $name string|ExpressionInterface column name

**Comment by rob006** [comment]
@darkdef What cases? #17473 is already fixed by #17490 in a way I proposed (by fixing `SimpleConditionBuilder::build()`). It just does not make sense to pass non-string to `Connection::quoteTableName()`, so code should be adjusted to not do that. This PR makes `Connection::quoteTableName()` interface more twisted only to silently handle bugs in other places.

**Comment by samdark** [comment]
Yes, makes sense. Reverted. Should be fixed another way.

**Comment by darkdef** [comment]
ok. I understood

File: framework/db/Connection.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (rob006) [reply framework/db/Connection.php:929]: Valid syntax is:

```
@param string|ExpressionInterface $name column name
```

But this should not be changed in the first place.

**Pull request merged**